types/tx_msg.go

```diff
@@ -65,8 +72,29 @@ func NewStdTx(msg Msg, sigs []StdSignature) StdTx {
 
 //nolint
 func (tx StdTx) GetMsg() Msg                   { return tx.Msg }
-func (tx StdTx) GetFeePayer() crypto.Address   { return tx.Signatures[0].PubKey.Address() } // XXX but PubKey is optional!
+func (tx StdTx) GetFeePayer() crypto.Address   { return tx.GetSigners()[0] }
 func (tx StdTx) GetSignatures() []StdSignature { return tx.Signatures }
+func (tx StdTx) GetSequence() int64            { return tx.Sequence }
+
+func CanonicalSignBytes(ctx Context, msg Msg) []byte {
+	obj := SignBytesObject{
+		ChainID:  "", // TODO
+		Sequence: 0,  // TODO
+		Msg:      msg,
+	}
+	// XXX: ensure some canonical form
+	b, err := json.Marshal(obj)
+	if err != nil {
+		panic(err)
+	}
+	return b
+}
+
+type SignBytesObject struct {
```

Comment:
Lets call this StdSignDoc, and have type signature be: 

```golang
type StdSignDoc struct {
  ChainID string,
  Nonce int64,
  MsgBytes []byte,
  AltBytes []byte,
}
```

precisely, and also include AltBytes in StdSignature!  That way the wallet client can store some info in there, and it will eventually get pruned away (since it isn't meant to affect the app state).

Reply:
Then use go-wire MarshalJSON to create the ultimate bytes to sign using the PrivKey.
Or here, we could use MarshalBinary, it might make more sense, since we don't need JSON anymore... we want JSON for Msg.GetSignBytes() because we need to show information to the user in the UX when signing.  But StdSignDoc would just be supported natively in `ckeystore`, we don't need the JSON structure.

So personally I would vote for marshalling StdSignDoc with go-wire.MarshalBinary().

In any case, we cannot leave it as `Msg Msg` for this purpose.

Reply:
What is an actual use case for the additional data bytes?  I feel like OP_RETURN is used in Bitcoin mostly just to add additional functionality that Bitcoin refuses to implement natively.  We should be adding features natively in protocol rather than giving people a mechanism to add second layer stuff using the data field.

Reply:
@jaekwon What is an actual use case for the additional data bytes? I feel like OP_RETURN is used in Bitcoin mostly just to add additional functionality that Bitcoin refuses to implement natively. We should be adding features natively in protocol rather than giving people a mechanism to add second layer stuff using the data field.